v2/terraform_flannel.py

```diff
@@ -250,32 +253,50 @@ def _copyFrom(self, src, dest, machine, windows=False, root=False):
             self.logging.error("Ansible failed to fetch file from %s with error: %s" % (machine, out))
             raise Exception("Ansible failed to fetch file from %s with error: %s" % (machine, out))
 
+    # Used by _runRemoteCmd to add more retries
+    def _parseAnsibleOutput(self, out):
+        json_data = json.loads(out)
+        machinesToRetry = []
+        for vm in json_data['stats'].keys():
+            if json_data['stats'][vm]['failures'] == 1 or json_data['stats'][vm]['unreachable'] == 1:
+                machinesToRetry.append(vm)
+        return machinesToRetry
+
     def _runRemoteCmd(self, command, machines, windows=False, root=False):
         self.logging.info("Running cmd %s on remote machines %s." % (command, machines))
-        cmd=["ansible"]
-        if root:
-            cmd.append("--become")
-        if windows:
-            task = "win_shell"
-        else:
-            task = "shell"
-            cmd.append("--key-file=%s" % self.opts.ssh_private_key_path)
-
-        cmd.append("'%s'" % " ".join(machines))
-        cmd.append("-m")
-        cmd.append(task)
-        cmd.append("-a")
-        cmd.append("'%s'" % command)
 
-        ret, out = self._waitForConnection(machines, windows=windows)
-        if ret != 0:
-            self.logging.error("No connection to machines. Error: %s" % out)
-            raise Exception("No connection to machines. Error: %s" % out)
+        def _runRemoteAnsible(command, machines):
+            cmd=["ansible"]
+            if root:
+                cmd.append("--become")
+            if windows:
+                task = "win_shell"
+            else:
+                task = "shell"
+                cmd.append("--key-file=%s" % self.opts.ssh_private_key_path)
+            cmd.append("'%s'" % " ".join(machines))
+            cmd.append("-m")
+            cmd.append(task)
+            cmd.append("-a")
+            cmd.append("'%s'" % command)
+            ret, out = self._waitForConnection(machines, windows=windows)
+            if ret != 0:
+                self.logging.error("No connection to machines. Error: %s" % out)
+            out, _, ret = utils.run_cmd(cmd, stdout=True, cwd=self.ansible_playbook_root, shell=True)
+            return ret, out, cmd
+
+        ret, out, cmd = _runRemoteAnsible(command, machines)
+        retries = self.opts.retries_ansible_adhoc
+        while retries !=0 and ret != 0:
+            retries -= 1
+            ret, out, cmd = _runRemoteAnsible(command, machines=self._parseAnsibleOutput(out))
+            self.logging.info("Ansible failed to run command %s. Retrying." % cmd)
```

Comment:
Minior nit, but this message should probably go before actually retring? Also, maybe a more useful message, like ansible failed to run on machines m with error e ?